lib/dfu_target_uart_host/dfu_target_uart_host.c

```diff
@@ -0,0 +1,319 @@
+					       context.to_write_len);
+		progress += context.to_write_len;
+		LOG_INF("Progress: %d %%", progress * 100 / context.file_size);
+		dfu_target_uart_host_send_error(
+			_DFU_UART_target_RemoteFnc_type_write,
+			err);
+		break;
+	case _DFU_UART_target_RemoteFnc_type_offset:
+		err = dfu_target_mcuboot_offset_get(&offs_get);
+		dfu_target_uart_host_send_offset(err, offs_get);
+		break;
+	case _DFU_UART_target_RemoteFnc_type_done:
+		dfu_target_uart_host_parse_done(fnc);
+		err = dfu_target_mcuboot_done(context.download_success);
+		if (err) {
+			dfu_target_uart_host_send_error(
+				_DFU_UART_target_RemoteFnc_type_done, err);
+			break;
+		}
+
+		err = dfu_target_mcuboot_schedule_update(
+			context.image_num);
+		dfu_target_uart_host_send_error(
+			_DFU_UART_target_RemoteFnc_type_done,
+			err);
+		break;
+	default:
+		break;
+	}
+}
+
+static void process_input_data(void)
+{
+	int err;
+	uint8_t *cursor;
+	size_t cursor_len;
+	size_t dec_len = 0;
+	struct DFU_UART_target_RemoteFnc fnc = {0};
+
+	while (true) {
+		cursor_len =
+			ring_buf_get_claim(&context.rx_rb, &cursor, UINT32_MAX);
+		if (cursor_len == 0) {
+			/* Let's reset it to prevent wrapping */
+			ring_buf_reset(&context.rx_rb);
+			return;
+		}
+
+		/* We try to put the whole buffer in the decode function.
+		 * If there is not enough data, the decode will throw an error.
+		 * If there is too much data, no error will be thrown and
+		 * the correct amount of decoded data will be given by the
+		 * decode function
+		 */
+		err = cbor_decode_DFU_UART_target_RemoteFnc(
+			cursor,
+			cursor_len,
+			&fnc,
+			&dec_len);
+		if (err) {
+			ring_buf_get_finish(&context.rx_rb, 0);
+			return;
+		}
+
+		execute_input_fnc(&fnc);
+
+		/* We need to remove the amount of consumed data */
+		err = ring_buf_get_finish(&context.rx_rb, dec_len);
+		if (err) {
+			LOG_ERR("ring_buf_get_finish returned %d", err);
+		}
+	}
+}
+
+noreturn static void uart_input_handler(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	while (1) {
+		k_sem_take(&context.lte_uart_sem, K_FOREVER);
+		process_input_data();
+	}
+}
+
+int dfu_target_uart_host_start(const struct device *uart)
+{
+	if (!uart) {
+		LOG_ERR("UART in not ready");
+		return -EINVAL;
+	}
+
+	context.uart_dev = uart;
+	context.out_buf_cursor = 0;
+
+	k_sem_init(&context.lte_uart_sem, 0, 1);
+	ring_buf_init(&context.rx_rb, ARRAY_SIZE(in_buf), in_buf);
+
+	if (!tid) {
+		tid = k_thread_create(&uart_handler_thread, uart_handler_stack,
+				      K_THREAD_STACK_SIZEOF(uart_handler_stack),
+				      uart_input_handler, NULL, NULL, NULL,
+				      CONFIG_DFU_TARGET_UART_HOST_HANDLER_PRIORITY,
+				      0,
+				      K_NO_WAIT);
+		k_thread_name_set(&uart_handler_thread, "dfu_uart");
+	}
+
+	uart_irq_callback_set(uart, uart_cb);
+	uart_irq_rx_enable(uart);
+
+	return 0;
+}
+
+int dfu_target_uart_host_stop(void)
+{
+	if (!context.uart_dev) {
+		return -EAGAIN;
+	}
```

Comment:
Multiple issues:

- No state checking, has this function been used before without being cleaned up?
- Semaphore re-initialised
- Ring buffer re-intiialised (and no cleanup)
- Context variables not cleaned up
- Thread does not get cleaned up/deleted when system is finished and persists
- If the UART device is not set i.e. system hasn't been setup, it returns error again, seems ENODEV is a more apt error code